src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixFourier.java

```diff
@@ -0,0 +1,309 @@
+
+		int rows = re.getNumRows();
+		int cols = re.getNumColumns();
+		if(!isPowerOfTwo(cols))
+			throw new RuntimeException("false dimensions");
+
+		fft(re.getDenseBlockValues(), im.getDenseBlockValues(), rows, cols, false);
+
+		return new MatrixBlock[] {re, im};
+	}
+
+	/**
+	 * Function to perform IFFT for each row of two given matrices. The first one represents the real values and the
+	 * second one the imaginary values. The output also contains one matrix for the real and one for the imaginary
+	 * values.
+	 *
+	 * @param re matrix object representing the real values
+	 * @param im matrix object representing the imaginary values
+	 * @return array of two matrix blocks
+	 */
+	public static MatrixBlock[] ifft_linearized(MatrixBlock re, MatrixBlock im) {
+
+		int rows = re.getNumRows();
+		int cols = re.getNumColumns();
+		if(!isPowerOfTwo(cols))
+			throw new RuntimeException("false dimensions");
+
+		ifft(re.getDenseBlockValues(), im.getDenseBlockValues(), rows, cols, false);
+
+		return new MatrixBlock[] {re, im};
+	}
+
+	/**
+	 * Function to perform FFT for two given double arrays. The first one represents the real values and the second one
+	 * the imaginary values. Both arrays get updated and contain the result.
+	 *
+	 * @param re          array representing the real values
+	 * @param im          array representing the imaginary values
+	 * @param rows        number of rows
+	 * @param cols        number of rows
+	 * @param inclColCalc if true, fft is also calculated for each column, otherwise only for each row
+	 */
+	public static void fft(double[] re, double[] im, int rows, int cols, boolean inclColCalc) {
+
+		double[] re_inter = new double[rows * cols];
+		double[] im_inter = new double[rows * cols];
+
+		for(int i = 0; i < rows; i++) {
+			fft_one_dim(re, im, re_inter, im_inter, i * cols, (i + 1) * cols, cols, 1);
+		}
+
+		if(inclColCalc) {
+			for(int j = 0; j < cols; j++) {
+				fft_one_dim(re, im, re_inter, im_inter, j, j + rows * cols, rows, cols);
+			}
+		}
+
+	}
+
+	/**
+	 * Function to perform IFFT for two given double arrays. The first one represents the real values and the second one
+	 * the imaginary values. Both arrays get updated and contain the result.
+	 *
+	 * @param re          array representing the real values
+	 * @param im          array representing the imaginary values
+	 * @param rows        number of rows
+	 * @param cols        number of rows
+	 * @param inclColCalc if true, fft is also calculated for each column, otherwise only for each row
+	 */
+	public static void ifft(double[] re, double[] im, int rows, int cols, boolean inclColCalc) {
+
+		double[] re_inter = new double[rows * cols];
+		double[] im_inter = new double[rows * cols];
+
+		if(inclColCalc) {
+			for(int j = 0; j < cols; j++) {
+				ifft_one_dim(re, im, re_inter, im_inter, j, j + rows * cols, rows, cols);
+			}
+		}
+
+		for(int i = 0; i < rows; i++) {
+			ifft_one_dim(re, im, re_inter, im_inter, i * cols, (i + 1) * cols, cols, 1);
```

Comment:
consider if this call can be parallel:

The code would be something along the lines of:

```
ExecutionContext pool = CommonThreadPool.get(k)
for( ...) {
pool.submit( () -> ifft_one_dim(re,im,re_inter, im_inter ...)
}
```

k is number of threads, and you can get that argument from CommonMath file.

Thanks

Reply:
full example of parallel execution: 
https://github.com/apache/systemds/blob/f4b7a47cbba9951210eef45263dc4a927b59225c/src/main/java/org/apache/sysds/runtime/frame/data/lib/FrameLibDetectSchema.java#L84